File: logic/src/engine.cpp
#include "command.h"
#include "cmdnum.h"

#include "cmdhandleinterface.h"
#include "cmdparseinterface.h"

#include "engine.h"

Engine::Engine(CmdParseInterface * _cmdp, const std::vector<CmdHandleInterface*>& _handlers) : cmdp(_cmdp){
    size_t s = _handlers.size();

    //Build command dispatch table
    if(s > 0) hndl_map[PUBLISH] = _handlers[0];
    if(s > 1) hndl_map[TIMELINE] = _handlers[1];
    if(s > 2) hndl_map[FOLLOW] = _handlers[2];
    if(s > 3) hndl_map[WALL] = _handlers[3];
}

std::vector<std::string> Engine::process(const std::string &input) {
    //Transform input command to command ID
    Command command = cmdp->parse(input);

    std::vector<std::string> ret;

    auto it = hndl_map.find(command.cmd);

    //Find command handle by ID and execute
    if(it != hndl_map.end()) {
        ret = it->second->handle(command);
    }

    return ret;
}

File: java/codechallenge/src/main/java/codechallenge/impl/Engine.java
package codechallenge.impl;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import codechallenge.interfaces.CmdHndlInterface;

public class Engine {

	private CmdParse cmdp;
	private Map<Integer, CmdHndlInterface> hndl_map;
	
	public Engine(CmdParse cmdp, Map<Integer, CmdHndlInterface> hndl_map) {
		super();
		this.cmdp = cmdp;
		this.hndl_map = hndl_map;
	}
	
	public List<String> process(final String input) {
		//Transform input command to command ID
		Command command = cmdp.parse(input);
		
		List<String> ret = new ArrayList<>();
		
		//Find command handle by ID and execute
		CmdHndlInterface hndl = hndl_map.getOrDefault(command.getCmd(), null);
		
		if(hndl != null) { ret = hndl.handle(command); }
		
		return ret;
	}
}
